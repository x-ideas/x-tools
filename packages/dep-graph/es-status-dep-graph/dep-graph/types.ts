import { ITokenExportedInfo } from '../exported/types';
import { ITokenImportedInfo } from '../imported/types';

// /**
//  * token的依赖信息
//  */
// export interface ITokenExportedDepInfo {
//   /** 导出的token的信息 */
//   // tokenInfo: ITokenExportedInfo;
//   /**
//    * 依赖这个token的文件路径集合
//    * 只考虑两个文件
//    */
//   // depBy: string[];
// }

/**
 *
 */
export interface IFileTokenDepInfo {
  [filePath: string]: (ITokenExportedInfo | ITokenImportedInfo)[];
}

export class TokenDepInfo {}
