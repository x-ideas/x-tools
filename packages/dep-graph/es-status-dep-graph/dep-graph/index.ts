export * from './token-dep-info';
