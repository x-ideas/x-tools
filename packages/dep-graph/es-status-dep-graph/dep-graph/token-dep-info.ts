import xmadge from '@x-tools/file-dep-graph';
import { findExportedInfoByFile } from '../exported';
import { findImportedInfoByFile } from '../imported';
import { IFileTokenDepInfo } from './types';

/**
 * 收集token的依赖信息
 * @export
 * @param {xmadge.MadgeInstance} madgeInst
 * @returns {IFileTokenDepInfo}
 */
export function collectTokenDepInfos(
  madgeInst: xmadge.MadgeInstance
): IFileTokenDepInfo {
  const depGraph: xmadge.MadgeModuleDependencyGraph = madgeInst.obj();

  const result: IFileTokenDepInfo = {};

  Object.keys(depGraph).forEach(path => {
    const importedTokens = findImportedInfoByFile(
      path,
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      madgeInst.config
    );

    const exportedTokens = findExportedInfoByFile(
      path,
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      madgeInst.config
    );

    result[path] = [...importedTokens, ...exportedTokens];
  });

  return result;
}
