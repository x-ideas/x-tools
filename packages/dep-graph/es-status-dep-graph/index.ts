import fs from 'fs';
import path from 'path';
import madge from 'madge';
import { parse, ParseResult } from '@babel/parser';
import traverse from '@babel/traverse';

/**
 * 找到文件的依赖图信息
 * @param {string} entryPath 入口文件路径
 * @throw
 */
async function findFileDependencyGraph(entryPath: string): Promise<madge.MadgeModuleDependencyGraph> {
  return madge(entryPath, {
    fileExtensions: ['js', 'ts', 'tsx', 'jsx'],
    tsConfig: '../../../tsconfig.json',
  }).then(res => res.obj());
}

// function analysisVariableDependencyGraph() {}

interface IExportedTokenInfo {
  /** 变量名字 */
  identifier: string;
  /** 命名/默认导出 */
  type: 'default' | 'named';
}

/**
 * 找到所有导出的接口的
 * 不同于typscript的interface的定义，这里把es语法中export导出的都看成是对外的接口
 * @example
 * ['a', 'o']
 * @export
 * @param {string} filePath
 * @returns
 */
export async function findAllExportedESInterfaces(filePath: string): Promise<IExportedTokenInfo[]> {
  const content = fs.readFileSync(filePath, {
    encoding: 'utf8',
  });
  const ast = parse(content, {
    sourceType: 'module',
  });
  const tokens: IExportedTokenInfo[] = [];
  tokens.push(...findExportNamedESInterfaces(ast));
  tokens.push(...findExportDefaultESInterfaces(ast));

  return tokens;
}

/**
 * 找到命名导出的变量
 * @param ast
 */
function findExportNamedESInterfaces(ast: ParseResult<babel.types.File>): IExportedTokenInfo[] {
  const tokens: IExportedTokenInfo[] = [];

  traverse(ast, {
    ExportNamedDeclaration(nodePath) {
      if (nodePath.node.declaration) {
        const { declaration } = nodePath.node;
        /**
         * export const a = '';
         */
        if (declaration.type === 'VariableDeclaration') {
          const { declarations } = declaration;
          declarations.forEach(declarator => {
            if (declarator.id.type === 'Identifier') {
              if (declarator)
                tokens.push({
                  identifier: declarator.id.name,
                  type: 'named',
                });
            }
          });
        }

        /**
         * export function f1 () {}
         */
        if (declaration.type === 'FunctionDeclaration') {
          if (declaration.id?.type === 'Identifier') {
            tokens.push({
              identifier: declaration.id.name,
              type: 'named',
            });
          }
        }
      }

      /**
       * export {
       *  c [as d]
       * }
       */
      nodePath.node.specifiers.forEach(specifier => {
        specifier.exported.type === 'Identifier' &&
          tokens.push({
            identifier: specifier.exported.name,
            type: 'named',
          });
      });
    },
  });

  return tokens;
}

/**
 * 找到默认导出的变量
 * @param ast
 */
function findExportDefaultESInterfaces(ast: ParseResult<babel.types.File>): IExportedTokenInfo[] {
  const tokens: IExportedTokenInfo[] = [];

  traverse(ast, {
    ExportDefaultDeclaration(nodePath) {
      /**
       * export default a;
       */
      if (nodePath.node.declaration.type === 'Identifier') {
        tokens.push({
          identifier: nodePath.node.declaration.name,
          type: 'default',
        });
      }

      /**
       * export default {
       *  a: '',
       * }
       */
      if (nodePath.node.declaration.type === 'ObjectExpression') {
        tokens.push({
          identifier: 'default',
          type: 'default',
        });
      }

      /**
       * export default c = ''
       */
      if (nodePath.node.declaration.type === 'AssignmentExpression') {
        nodePath.node.declaration.left.type === 'Identifier' &&
          tokens.push({
            identifier: nodePath.node.declaration.left.name,
            type: 'default',
          });
      }
    },
  });

  return tokens;
}

// export async function run(entryPath: string) {
//   const moduleDepGraph = await findFileDependencyGraph(entryPath);

//   Object.entries(moduleDepGraph).forEach(([moduleFile, moduleDepFiles]) => {});
// }
