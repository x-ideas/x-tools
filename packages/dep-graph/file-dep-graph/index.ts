import xmadge from 'madge';

export default xmadge;
