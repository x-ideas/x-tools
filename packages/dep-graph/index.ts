import fs from 'fs';
import path from 'path';
import { parse } from '@babel/parser';
import traverse from '@babel/traverse';

function run() {
  const file = path.resolve(__dirname, './demo.ts');
  const content = fs.readFileSync(file, {
    encoding: 'utf8',
  });

  // process.stdout.write(contenxt);
  const result = parse(content, {
    sourceType: 'module',
  });
  traverse(result, {
    enter(node) {
      process.stdout.write(node.node.type);
      process.stdout.write('\n');
    },
  });
}

run();
