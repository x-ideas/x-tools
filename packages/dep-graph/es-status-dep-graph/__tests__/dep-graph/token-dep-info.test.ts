import xmadge from '@x-tools/file-dep-graph';
import path from 'path';
import { collectTokenDepInfos } from '../../dep-graph';

describe('token-dep-info', () => {
  it('circular dep', async () => {
    const entryPath = path.resolve(
      __dirname,
      '../__testfixtures__/demo-proj/index.ts'
    );
    const inst = await xmadge(entryPath, {
      tsConfig: path.resolve(__dirname, '../__testfixtures__/tsconfig.json'),
      fileExtensions: ['js', 'ts', 'jsx', 'tsx', 'mjs'],
      // NOTE: 保证返回的是相对当前的路径
      baseDir: './',
    });

    const tokenDepInfos = collectTokenDepInfos(inst);

    expect(tokenDepInfos).toMatchSnapshot();
  });
});
