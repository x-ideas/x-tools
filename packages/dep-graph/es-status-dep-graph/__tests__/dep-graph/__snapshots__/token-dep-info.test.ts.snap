// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`token-dep-info circular dep 1`] = `
Object {
  "packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-a.ts": Array [
    Object {
      "exportKind": "value",
      "exported": "depA",
      "express": "export const depA = '';",
      "filePath": "packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-a.ts",
      "local": "depA",
      "sourcePath": "packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-a.ts",
      "type": "named",
    },
  ],
  "packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-b.ts": Array [
    Object {
      "express": "import _ from 'lodash';",
      "filePath": "packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-b.ts",
      "importKind": "value",
      "imported": "default",
      "local": "_",
      "sourcePath": "/Users/xjx/Documents/programs/myself/x-tools/node_modules/.pnpm/lodash@4.17.21/node_modules/lodash/lodash.js",
      "type": "Default",
    },
    Object {
      "express": "import { TDepC } from './dep-c';",
      "filePath": "packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-b.ts",
      "importKind": "value",
      "imported": "TDepC",
      "local": "TDepC",
      "sourcePath": "/Users/xjx/Documents/programs/myself/x-tools/packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-c.ts",
      "type": "Named",
    },
    Object {
      "exportKind": "type",
      "exported": "IDepB",
      "express": "export interface IDepB {
  name: TDepC;
}",
      "filePath": "packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-b.ts",
      "local": "IDepB",
      "sourcePath": "packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-b.ts",
      "type": "named",
    },
    Object {
      "exportKind": "value",
      "exported": "omit",
      "express": "export const omit = _.omit;",
      "filePath": "packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-b.ts",
      "local": "omit",
      "sourcePath": "packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-b.ts",
      "type": "named",
    },
  ],
  "packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-c.ts": Array [
    Object {
      "express": "import { Hello } from 'dep-d';",
      "filePath": "packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-c.ts",
      "importKind": "value",
      "imported": "Hello",
      "local": "Hello",
      "sourcePath": "/Users/xjx/Documents/programs/myself/x-tools/packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-d.ts",
      "type": "Named",
    },
    Object {
      "exportKind": "type",
      "exported": "TDepC",
      "express": "export type TDepC = string;",
      "filePath": "packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-c.ts",
      "local": "TDepC",
      "sourcePath": "packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-c.ts",
      "type": "named",
    },
  ],
  "packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-d.ts": Array [
    Object {
      "express": "import { omit } from 'dep-b';",
      "filePath": "packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-d.ts",
      "importKind": "value",
      "imported": "omit",
      "local": "omit",
      "sourcePath": "/Users/xjx/Documents/programs/myself/x-tools/packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-b.ts",
      "type": "Named",
    },
    Object {
      "exportKind": "value",
      "exported": "DepD",
      "express": "export const DepD = 'DepD';",
      "filePath": "packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-d.ts",
      "local": "DepD",
      "sourcePath": "packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-d.ts",
      "type": "named",
    },
    Object {
      "exportKind": "value",
      "exported": "ETarget",
      "express": "export enum ETarget {}",
      "filePath": "packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-d.ts",
      "local": "ETarget",
      "sourcePath": "packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-d.ts",
      "type": "named",
    },
    Object {
      "exportKind": "value",
      "exported": "Hello",
      "express": "export function Hello(): string {
  return 'Hello';
}",
      "filePath": "packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-d.ts",
      "local": "Hello",
      "sourcePath": "packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-d.ts",
      "type": "named",
    },
  ],
  "packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/index.ts": Array [
    Object {
      "express": "import * as A from './dep-a';",
      "filePath": "packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/index.ts",
      "importKind": "value",
      "imported": undefined,
      "local": "A",
      "sourcePath": "/Users/xjx/Documents/programs/myself/x-tools/packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-a.ts",
      "type": "Namespaces",
    },
    Object {
      "express": "import { omit } from './dep-b';",
      "filePath": "packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/index.ts",
      "importKind": "value",
      "imported": "omit",
      "local": "omit",
      "sourcePath": "/Users/xjx/Documents/programs/myself/x-tools/packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-b.ts",
      "type": "Named",
    },
    Object {
      "express": "import run from './dep-c';",
      "filePath": "packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/index.ts",
      "importKind": "value",
      "imported": "default",
      "local": "run",
      "sourcePath": "/Users/xjx/Documents/programs/myself/x-tools/packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-c.ts",
      "type": "Default",
    },
    Object {
      "express": "import { Hello } from './dep-d';",
      "filePath": "packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/index.ts",
      "importKind": "value",
      "imported": "Hello",
      "local": "Hello",
      "sourcePath": "/Users/xjx/Documents/programs/myself/x-tools/packages/dep-graph/es-status-dep-graph/__tests__/__testfixtures__/demo-proj/dep-d.ts",
      "type": "Named",
    },
  ],
}
`;
